### ISSUE TITLE
Do a better job of serializing non-error objects

### ISSUE THREAD

**Issue by birtles** [birchill/bugsnag-zero#110]
See https://github.com/birchill/bugsnag-zero/issues/109#issuecomment-1632021200 onwards

**Comment by xPaw** [comment]
Perhaps `InvalidError` could be wrapped into its own plugin that can be optionally included somehow?

When not included, it would not report such errors.

**Comment by birtles** [comment]
> Perhaps `InvalidError` could be wrapped into its own plugin that can be optionally included somehow?
> 
> When not included, it would not report such errors.

Yeah, I was thinking about that too. I started working on this over in the [serialize-error-objects](https://github.com/birchill/bugsnag-zero/tree/serialize-error-objects) branch and so far the code's pretty small. When it's ready, I'll measure the size difference and see if it deserves its own plugin or not.

**Comment by birtles** [comment]
> Just to add to this, `String(obj)` can throw `Cannot convert object to primitive value`. _Don't you just love random errors coming from user script and extensions?_
> 
> Repro: `String(Object.create(null))`

That's good to know. So if we make this serialization a plugin based and our fallback is `String()` we'll need to wrap it in `try ... catch`.

**Comment by birtles** [comment]
Made some progress on the serialization. Currently working on more useful and robust serialization for `Map`s and `Set`s. With that added complexity, it seems pretty likely we'll want to make this an optional plugin rather than a core feature.

**Comment by birtles** [comment]
The serialization is probably good enough for now. However, it might be nice to apply it to more than just unrecognized error objects. Potentially anything in the metadata would benefit from proper serialization that takes care of cyclic dependencies, exotic objects etc. I noticed that's what the official client does.

That said, this serialization currently takes an extra 7Kb or so (uncompressed).

I've also noticed a bunch of other places where we use `JSON.stringify` without considering if it throws, like in the `redact-keys` plugin.

I think we could probably make a few changes:

- In general, feature-detect `structuredClone` and use that in place of `JSON.parse(JSON.stringify(object))` when available
- Make the redact keys plugin detect cycles
- Create a plugin that provides better serialization we use to serialize the whole payload
- When the plugin is not available, use `JSON.stringify` and, if it throws, drop the metadata and try serializing again since it's likely something in the metadata to blame

That would bring us closer to the official plugin whilst still making the serialization code and redaction code opt-in.

The other change I might make is that I wrote the serialization code based on `fast-safe-stringify` which tries to optimize performance by modifying the object in place. However, we're not really that concerned with serialization performance. Also, I don't think the code handles getters that throw. In both regards I think the [`safe-json-stringify` library](https://github.com/bugsnag/safe-json-stringify/) used by the official class is better.

Working on a copy of the object will be slower but it's probably less code since we don't need to restore changes to the original object afterwards.
